Create OpenAPIV3 schema for secret gen controller package and generate package with the schema

shivaani0505 opened the pull request:
## What this PR does / why we need it
Create OpenAPIV3 schema for secret gen controller package and generate package with the schema

- [X] Add make target and corresponding script for generating open API schema for package 
- [X] Modify make target for make push-package and corresponding script for verifying the correct schema exists in package before imgpkg  bundle is pushed
- [X] Add overlay for package for openAPI schema to remove package-template.yaml
- [X] Modify the doc located at docs/site/content/docs/latest/designs/package-process.md to add the required steps
- [X] Add logic for storing the generated contents for testing schema in artifacts 
<img width="300" alt="artifacts_directory" src="https://user-images.githubusercontent.com/57377053/148591197-9dcf8a60-bb84-4a4e-a948-ed3ba6afd5ce.png">

## Details for the Release Notes (PLEASE PROVIDE)
<!--
Unless this is a trivial change, we want to know more about your contribution!
This can even be a TLDR version of the "What this PR does".
If a trivial change, just write "NONE" in the release-note block below.
Otherwise, a release note is required:
-->
```release-note
Create OpenAPIV3 schema for secret gen controller package and generate package with the schema
```

## Which issue(s) this PR fixes
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes: #2718

## Describe testing done for PR
Tested for both type of packages which contain sample values and one which don't.
Assumption: sample-values directory exists under bundle directory
Sample package tested: **secretgen-controller**
**Step 1:** Use make target for generating openAPI v3 schema

`make generate-openapischema-package PACKAGE=secretgen-controller VERSION=0.7.1`

**Step 2:**

`make push-package PACKAGE=secretgen-controller VERSION=0.7.1 TAG=0.7.1 `

Once the package is pushed successfully, copy the SHA and paste it in package.yaml at the appropriate location
Use the generated files for creating a PR.

## Special notes for your reviewer

**Note:** Doc to be updated with above make targets once initial review of scripts is done

All the above TKRs refer to version 0.7.1 of secretgen-controller , because all the current TKRs refer to that version

shivaani0505 committed:
Create OpenAPIV3 schema for secret gen controller package and generate package with the schema

danniel1205 commented on addons/packages/secretgen-controller/0.7.1/package.yaml:
```diff
@@ -3,23 +3,41 @@ kind: Package
     spec:
       fetch:
-        - imgpkgBundle:
-            image: projects.registry.vmware.com/tce/secretgen-controller@sha256:4248e36490eb888d7f8bd0b62739a5acc3f178f67d8c2abfb3a6181b814c074e
```
Question: If a maintainer uses the `package-template.yaml` above, will the SHA always be the one specified? If so, it is an error-prone process because maintainer could easily forget about updating it. That makes me think whether the package folder becomes cleaner if we only check in the changes from `schema.yaml` and `package.yaml` (that also helps remove some duplications). And provide a doc on how to add/update schema to package.

shivaani0505 replied:
Since we are adding new files like schema.yaml to the package , we need to run `make push-package` which generates new SHA for package.yaml. If the SHA is being updated in package.yaml, the question is person can forget to add SHA  in package-template.
I was thinking of adding a doc and adding some make commands to help in schema and package generation. What should be the location of doc?

shivaani0505 committed:
Fix lint issue

Signed-off-by: Shivaani Gupta <gshivaani@vmware.com>

vijaykatam commented on addons/packages/secretgen-controller/0.7.1/package.yaml:
```diff
@@ -3,23 +3,41 @@ kind: Package
+  valuesSchema:
+    openAPIv3:
+      type: object
+      additionalProperties: false
```
Could we add `description` here? See example https://github.com/kubernetes-sigs/cluster-api/blob/main/config/crd/bases/cluster.x-k8s.io_clusters.yaml#L30

shivaani0505 replied:
yes that can be added

vijaykatam commented on addons/packages/secretgen-controller/0.7.1/package.yaml:
```diff
@@ -3,23 +3,41 @@ kind: Package
+      additionalProperties: false
+      properties:
+        secretgenController:
+          type: object
```
For properties of type object, we'll need to add `default: {}` so that openAPI utilities can use this to generate an object out of this.

shivaani0505 replied:
@vijaykatam , Did you mean something like below, check out default:
```
openapi: 3.0.0
info:
  version: 0.1.0
  title: Schema for data values, generated by ytt
paths: {}
components:
  schemas:
    dataValues:
      type: object
      additionalProperties: false
      description: OpenAPIV3 Schema for secret gen controller
      properties:
        secretgenController:
          type: object
          additionalProperties: false
          description: Configuration for secret gen controller
          properties:
            namespace:
              type: string
              default: secretgen-controller
              description: Namespace for secret gen controller
            createNamespace:
              type: boolean
              default: false
              description: Whether to create namespace for secret gen controller if not present
          default: {}
```

shivaani0505 replied:
I ask because default and properties is mutually exclusive according to carvel team and above is not valid schema. 
If something else is expected, let me know or if there is an example which has default for object type

vijaykatam replied:
I haven't looked at openapi schema spec(https://spec.openapis.org/oas/v3.1.0) to validate if what you are saying is true but the reason I brought it up is because k8s apiextensions-server defaulting relies on non-nil default for an object, see https://github.com/kubernetes/apiextensions-apiserver/tree/master/pkg/apiserver/schema/defaulting

blc1996 replied:
Hi @shivaani0505 , please refer to the changes in https://github.com/vmware-tanzu/tanzu-framework/pull/1422/files#diff-8b8b2f2c353cbd029665d4e29754f2ed8d12ab749d2968431ed6563f113a01e4R47. It seems that Vijay handled this case in the util func and we don't need to add empty default objects.

shivaani0505 replied:
Sure @blc1996 . The test is blocking this MR. I have 2 options for testing listed above. We can discuss on those or a completely new one for testing

shivaani0505 replied:
Added all the testing tasks as discussed, so now closing this

vijaykatam reviewed (commented):
What is the plan to test the schema?

shivaani0505 commented on addons/packages/secretgen-controller/0.7.1/bundle/config/openapi-schema.yaml:
```diff
@@ -0,0 +1,24 @@
+openapi: 3.0.0
```
I can add dev doc. Is this location fine for the doc => https://github.com/vmware-tanzu/community-edition/tree/main/docs

shivaani0505 replied:
Added dev doc under community-edition/addons/packages/

shivaani0505 replied:
Changed the location of doc to docs/site/content/docs/latest/designs/package-process.md as suggested by @seemiller

shivaani0505 committed:
Add dev doc for generating openAPI v3 schema and address review comments

Signed-off-by: Shivaani Gupta <gshivaani@vmware.com>

shivaani0505 committed:
Fix markdown errors

Signed-off-by: Shivaani Gupta <gshivaani@vmware.com>

seemiller reviewed (changes requested):
We need to find a better place for this documentation to live. Perhaps work it into docs/site/content/docs/latest/designs/package-process.md ?

shivaani0505 commented:
> We need to find a better place for this documentation to live. Perhaps work it into docs/site/content/docs/latest/designs/package-process.md ?

I hv asked the same ques. For now I put it in addons/packages, but can change to location which is preferred

shivaani0505 committed:
Working the doc for OpenAPI schema into package-process.md

Signed-off-by: Shivaani Gupta <gshivaani@vmware.com>

shivaani0505 commented:
> We need to find a better place for this documentation to live. Perhaps work it into docs/site/content/docs/latest/designs/package-process.md ?

Incorporated as a part of docs/site/content/docs/latest/designs/package-process.md

shivaani0505 committed:
Adding missing configuration for secret gen controller in README and updating relevant files

Signed-off-by: Shivaani Gupta <gshivaani@vmware.com>

dvonthenen commented:
just trying to follow up... is this ready to merge or is there more work needed?

blc1996 commented on addons/packages/secretgen-controller/0.7.1/package.yaml:
```diff
@@ -3,23 +3,42 @@ kind: Package
 metadata:
   name: secretgen-controller.community.tanzu.vmware.com.0.7.1
 spec:
+  valuesSchema:
```
Since this part is generated out of the `schema.yaml` of each package, can we do some automation so that we can avoid manually copy and paste?

shivaani0505 replied:
@blc1996 Step 5 command in this https://github.com/vmware-tanzu/community-edition/blob/732e5cd301bda4b3d9a2e22c1122d43fbd886428/docs/site/content/docs/latest/designs/package-process.md#8-generate-openapiv3-schema-and-embed-it-in-a-package, is used to generate the above section so it is automated and not manually added

shivaani0505 commented:
> just trying to follow up... is this ready to merge or is there more work needed?

@dvonthenen This PR has the changes but I am figuring out to add the tests for schema

shivaani0505 committed:
Fix command in documentation for generating open api schemal

shivaani0505 commented:
> What is the plan to test the schema?
@vijaykatam , Plan for testing is:
**Way 1:**
Perform Step 2 and Step 3 to verify schema in https://github.com/vmware-tanzu/community-edition/blob/732e5cd301bda4b3d9a2e22c1122d43fbd886428/docs/site/content/docs/latest/designs/package-process.md#8-generate-openapiv3-schema-and-embed-it-in-a-package, to generate expected yaml
```
ytt -f values.yaml -f schema.yaml -f template.yaml
ytt -f schema.yaml --data-values-schema-inspect -o openapi-v3 > openapi-schema.yaml
```
Here, let's say we have a key which has been declared as a boolean in schema and values.yaml gives a string or something else, error is thrown and no output is generated. So all those test cases which don't abide by the declared schema would throw error.

We can have make target for running that command.

**Way 2:**
Introduce command options like  --data-values-schema-inspect in the below tests. Haven't run these tests in TCE but we might be able to extend these 
https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/test/pkg/ytt/ytt.go

We can discuss more on the above 2 options or if you think there should be some other way of testing, we can go with that.

blc1996 commented:
> > What is the plan to test the schema?
> > @vijaykatam , Plan for testing is:
> > **Way 1:**
> > Perform Step 2 and Step 3 to verify schema in https://github.com/vmware-tanzu/community-edition/blob/732e5cd301bda4b3d9a2e22c1122d43fbd886428/docs/site/content/docs/latest/designs/package-process.md#8-generate-openapiv3-schema-and-embed-it-in-a-package, to generate expected yaml
> 
> ```
> ytt -f values.yaml -f schema.yaml -f template.yaml
> ytt -f schema.yaml --data-values-schema-inspect -o openapi-v3 > openapi-schema.yaml
> ```
> 
> Here, let's say we have a key which has been declared as a boolean in schema and values.yaml gives a string or something else, error is thrown and no output is generated. So all those test cases which don't abide by the declared schema would throw error.
> 
> We can have make target for running that command.
> 
> **Way 2:** Introduce command options like --data-values-schema-inspect in the below tests. Haven't run these tests in TCE but we might be able to extend these https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/test/pkg/ytt/ytt.go
> 
> We can discuss more on the above 2 options or if you think there should be some other way of testing, we can go with that.

It would be something similar to the template tests in tkg-packages repo, but without checking the actual content of the generated manifests. We just check if YTT can successfully render the manifests with the give `schema.yaml` and `values.yaml`. Ideally, it should be run as a pipeline in TCE automatically. One thing missing here is the sample values we have in tkg-packages for some packages.

blc1996 commented on docs/site/content/docs/latest/designs/package-process.md:
```diff
@@ -454,7 +454,133 @@ spec:
+
+   `ytt -f schema.yaml --data-values-schema-inspect -o openapi-v3 > openapi-schema.yaml`
+
+4. Create package-template.yaml file from package.yaml by adding the following 2 lines in `spec` field:
```
We should have done this when we create the schema of each package and put the `package-template.yaml` file inside the bundle. The developers don't need to worry about this step later on.

blc1996 replied:
Can we update this README as we don't need to create template anymore

shivaani0505 replied:
Yes , I am updating it now and also fixing lint issues @blc1996 
I had put a note above that I will change it once initial review for make targets is done.

shivaani0505 replied:
Done

blc1996 commented on docs/site/content/docs/latest/designs/package-process.md:
```diff
@@ -454,7 +454,133 @@ spec:
+
+5. You can use this OpenAPI schema as documentation, or use it in a Package, to document what inputs are allowed in the package.
+
+   `ytt -f ../../package-template.yaml --data-value-file openapi=openapi-schema.yaml > ../../package.yaml`
```
We should probably make this command part of a make target, like integrate with `make push-package`?

shivaani0505 replied:
We are now checking the generated schema as a part of make push-package and we have a separate make target for generating openAPI schema

shivaani0505 committed:
Fix documentation and comments

shivaani0505 committed:
Add make target and corresponding script for generating open API schema for package
Modify make target for make push-package and corresponding script for verifying the correct schema exists in package before imgpkg  bundle is pushed
Add overlay for package for openAPI schema to remove package-template.yaml

shivaani0505 committed:
Fix lint issues for shell

shivaani0505 committed:
Update the doc package-process.md with the latest steps for generating openAPIv3  schema

blc1996 commented on Makefile:
```diff
@@ -353,9 +353,19 @@ lock-package-images: check-carvel # Updates the image lock file for a package. U
+	&& ytt -f schema.yaml --data-values-schema-inspect -o openapi-v3 > openapi-schema.yaml \
+	&& ytt -f ../../package.yaml -f ../../../../package-overlay/package-overlay.yaml --data-value-file openapi=openapi-schema.yaml > ../../generated-package.yaml \
+	&& mv ../../generated-package.yaml ../../package.yaml \
+	&& rm openapi-schema.yaml
```
Can we have some log message to tell the user something like `package.yaml` has been updated with the latest `schema.yaml` in its valuesSchema field.

shivaani0505 replied:
Done

maralavi commented on hack/packages/check-sample-values-package.sh:
```diff
@@ -0,0 +1,49 @@
+# Copyright 2022 VMware Tanzu Community Edition contributors. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+# this shell script checks the existence os a sample values file for a Tanzu package and verifies if ytt templates are generated.
```
typo "os a" --> of

shivaani0505 replied:
Fixed

blc1996 commented on hack/packages/check-sample-values-package.sh:
```diff
@@ -0,0 +1,49 @@
+	${yttCmd}
+}
+
+check_sample_values_existence_for_package
```
I feel that this function name can be improved a bit. It not only checks the existence of sample values file, but also renders ytt manifests.

shivaani0505 replied:
Changing function name to check_sample_values_and_render_ytt

shivaani0505 replied:
Fixed

maralavi commented on hack/packages/check-sample-values-package.sh:
```diff
@@ -0,0 +1,49 @@
+
+ROOT_DIR="addons/packages"
+BUNDLE_DIR="bundle"
+CONFIG_DIR="config"
```
These definitions for ```BUNDLE_DIR="bundle"``` and ```CONFIG_DIR="config"``` seem redundant.

shivaani0505 replied:
Fixed

maralavi replied:
@shivaani0505 I think you can remove these for ```verify-openapischema-for-package.sh``` as well

shivaani0505 replied:
Fixed

maralavi commented on hack/packages/check-sample-values-package.sh:
```diff
@@ -0,0 +1,49 @@
+    yttCmd="${yttCmd} -f ../sample-values/*.yaml"
+  fi
+  cd "${CONFIG_DIR}" || exit
+	${yttCmd}
```
if sample values existed, as ```yttCmd``` is replaced in ```yttCmd="${yttCmd} -f ../sample-values/*.yaml"```, can it still successfully run line 43?

shivaani0505 replied:
Yes it does. I tested it with sample values

shivaani0505 commented:
@seemiller , I am using some binaries like yq and kapp. As of now I do not see hack/tools in TCE directory being used to download binaries.
Would it be fine if I introduce that similar to our downstream repo. wanted to know if TCE is fine with this change 
Otherwise, if users don't have the binaries some scripts will fail like (https://github.com/vmware-tanzu/community-edition/blob/0b7d33e5bccf067c403f7f52e28c181c8730f567/hack/packages/verify-openapischema-for-package.sh)
If not, then we can add log messages in our scripts to ask users to install those binaries beforehand.

cc: @vijaykatam

shivaani0505 committed:
Fix review comments

shivaani0505 committed:
Address review comments

shivaani0505 commented on hack/packages/verify-openapischema-for-package.sh:
```diff
@@ -0,0 +1,75 @@
+		return 0
+	else
+		echo -e "${RED}Error: Found diff in ${YELLOW}$file1 ${RED}and ${YELLOW}$file2${NC}"
+		echo -e "${RED}Please examine both files${NC}"
```
Added.

shivaani0505 commented on hack/packages/verify-openapischema-for-package.sh:
```diff
@@ -0,0 +1,75 @@
+  # Get rid of all the output lines that say there is no difference.
+  # Get rid of the output line that says "Succeeded".
+  # Expect that there are no output lines leftover in the output.
+  if eval "$(kapp tools diff --file "$file1" --file2 "$file2" --changes --json \
```
file1 and file2 have been renamed.
For generation of files, will use artifacts directory. Changes to follow soon.

blc1996 commented on Makefile:
```diff
@@ -353,9 +353,20 @@ lock-package-images: check-carvel # Updates the image lock file for a package. U
+	@printf "\n===> generating OpenAPIv3 schema for $${PACKAGE}/$${VERSION}\n";\
+	./hack/packages/check-sample-values-and-render-ytt.sh $(PACKAGE) $(VERSION) \
+	&& cd addons/packages/$${PACKAGE}/$${VERSION}/bundle/config \
+	&& ytt -f schema.yaml --data-values-schema-inspect -o openapi-v3 > openapi-schema.yaml \
```
Can we put this `openapi-schema.yaml` in `artifacts/<package_name>/<version>` as well?

shivaani0505 replied:
Fixed

blc1996 commented on addons/packages/package-overlay/package-overlay.yaml:
```diff
@@ -0,0 +1,9 @@
+spec:
+#@overlay/match missing_ok=True
+  valuesSchema:
+    openAPIv3:  #@ yaml.decode(data.values.openapi)["components"]["schemas"]["dataValues"]
```
Should have ytt annotation like `@overlay/replace `. Otherwise, it will keep overlaying on the existing pakcage.yaml valuesSchema contents.

shivaani0505 replied:
Fixed

shivaani0505 committed:
Add logic for storing the generated contents for testing schema in artifacts and fixing review comments

shivaani0505 committed:
Add check in verify schema script only for packages which have schema.yaml

blc1996 reviewed (dismissed):
LGTM

dvonthenen commented on Makefile:
```diff
@@ -353,9 +353,21 @@ lock-package-images: check-carvel # Updates the image lock file for a package. U
+	./hack/packages/verify-openapischema-for-package.sh $(PACKAGE) $(VERSION) \
+	&& cd addons/packages/$${PACKAGE}/$${VERSION} && imgpkg push --bundle $(OCI_REGISTRY)/$${PACKAGE}:$${VERSION} --file bundle/;\
+
+generate-openapischema-package: #Generate package with OpenAPI v3 schema
```
I don't really know much about how the packaging stuff works, so please bear with me. Is this `generate-openapischema-package` specific to a particular package? It doesn't seem like it since there is a `$${PACKAGE}` var out there, but I'm wondering if we should put any kind of package pre or post processing into a sub-Makefile located in the `/addons` or `/addons/packages` folder as not to bloat the upper level Makefile. Thoughts?

cc: @seemiller @jpmcb

shivaani0505 replied:
When I was adding make target, I was recommended by @seemiller to add it to the main Makefile and yes, it is used to generate openapischema-package specific to a particular package.
Also, make push package which is now modified will only look and verify  generated schema for packages which have schema.yaml file

dvonthenen replied:
my comment here is just more like just an open question to people who are better suited/qualified to answer. just wanted to bring awareness to try not bloat the main Makefile if possible. if this is the best or only place for this, then totally cool by me. adding @seemiller

seemiller replied:
@dvonthenen I like your idea of moving these to a more package specific makefile. However, could we do that as it's own separate issue?

shivaani0505 commented:
management cluster delete is failing, in the morning all checks were passing, @dvonthenen @seemiller , do you have any idea why this is happening

shivaani0505 committed:
Fixing grammatical errors

shivaani0505 commented on docs/site/content/docs/latest/designs/package-process.md:
```diff
@@ -454,7 +454,110 @@ spec:
+### 8. Generate openAPIv3 schema and embed it in a package
+
+Follow the below mentioned steps to get started on generating openAPIv3 schema and specifying it in a package.
+This process works for both kinds of package, one which have sample values defined like csi, cpi and also for ones which don't like secretgen-controller, kapp-controller to name a few.
```
Fixed

shivaani0505 commented on docs/site/content/docs/latest/designs/package-process.md:
```diff
@@ -454,7 +454,110 @@ spec:
+
+2. Generate OpenAPI v3 schema using the following make target
+
+   Let's use secretgen-controller package as an example to generate schema and embed it in Package
```
Fixed

shivaani0505 commented on docs/site/content/docs/latest/designs/package-process.md:
```diff
@@ -454,7 +454,110 @@ spec:
+3. You can now use `make push-package` for your package
+
+   This performs 2 steps:
+   * Verifies if the openAPIv3 schema embedded in package matches exactly with the openAPIv3 schema generated.It also prevents pushing package's imgpkg bundle without the openAPI schema embedded.
```
Fixed

vijaykatam commented on Makefile:
```diff
@@ -352,9 +352,21 @@ lock-package-images: check-carvel # Updates the image lock file for a package. U
+	&& cd addons/packages/$${PACKAGE}/$${VERSION} \
+	&& mkdir -p ${ARTIFACTS_DIR} \
+	&& cd ${ARTIFACTS_DIR} \
+	&& ytt -f ../bundle/config/schema.yaml --data-values-schema-inspect -o openapi-v3 > openapi-schema.yaml \
```
(question): Are devs required to ensure that they are using the right version of ytt and other binaries? Is there a doc somewhere that states the versions.

vijaykatam commented on hack/packages/verify-openapischema-for-package.sh:
```diff
@@ -0,0 +1,75 @@
+    mkdir -p "${ARTIFACTS_DIR}"
+    cd "${ARTIFACTS_DIR}" || exit
+    ytt -f ../bundle/config/schema.yaml --data-values-schema-inspect -o openapi-v3 > generated-openapi-schema.yaml
+    yq e '.components.schemas.dataValues' generated-openapi-schema.yaml > schema-contents.yaml
```
(comment): Since we introducing `yq` and `kapp` into the dev process it would be good to create tool targets for these that ensure install with appropriate versions (similar to downstream).

blc1996 replied:
Discussed with @shivaani0505 about this before. TCE doesn't have such make targets to install tools. They directly use the binaries installed in the users' path. We probably need TCE folks to chime in if we decide to add these tool make targets.

shivaani0505 replied:
I had suggested the same in the link https://github.com/vmware-tanzu/community-edition/pull/2719#issuecomment-1006976852, and also discussed with @seemiller . He said he would bring it up in TCE meeting

vijaykatam replied:
Sounds good.

vijaykatam reviewed (commented):
Minor comments. Lets follow up with a CI target for verifying that schema was added to Package CR.

blc1996 commented on hack/packages/check-sample-values-and-render-ytt.sh:
```diff
@@ -0,0 +1,50 @@
+GREEN='\033[0;32m'
+
+check_sample_values_and_render_ytt() {
+  sample_values_dir="${BUNDLE_DIR}/sample-values"
```
Just realized that if `sample-values` is put inside `bundle`, it will be pushed as package content when we execute `make push-package`. It is something we desire? If not, we need to move `sample-values` a level up. @shivaani0505 @vijaykatam

vijaykatam replied:
yeah I missed it as well, `sample-values` shouldn't be part of the bundle.

shivaani0505 replied:
will move this one level up
